lib/ruby_lsp/requests/path_completion.rb

```diff
@@ -0,0 +1,96 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # ![Path completion demo](../../misc/path_completion.gif)
+    #
+    # The [completion](https://microsoft.github.io/language-server-protocol/specification#textDocument_completion)
+    # request looks up Ruby files in the $LOAD_PATH to suggest path completion inside `require` statements.
+    #
+    # # Example
+    #
+    # ```ruby
+    # require "ruby_lsp/requests" # --> completion: suggests `base_request`, `code_actions`, ...
+    # ```
+    class PathCompletion < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, position: Document::PositionShape).void }
+      def initialize(document, position)
+        super(document)
+
+        @tree = T.let(Support::PrefixTree.new(collect_load_path_files), Support::PrefixTree)
+
+        char_position = document.create_scanner.find_char_position(position)
+        @target = T.let(find(T.must(document.tree), char_position), T.nilable(SyntaxTree::TStringContent))
+      end
+
+      sig { override.returns(T.all(T::Array[LanguageServer::Protocol::Interface::CompletionItem], Object)) }
+      def run
+        # no @target means the we are not inside a `require` call
+        return [] unless @target
+
+        text = @target.value
+        @tree.search(text).sort.map do |path|
+          build_completion(path, path.delete_prefix(text))
+        end
+      end
+
+      private
+
+      sig { returns(T::Array[String]) }
+      def collect_load_path_files
+        $LOAD_PATH.flat_map do |p|
+          Dir.glob("**/*.rb", base: p)
+        end.map! do |result|
+          result.delete_suffix!(".rb")
+        end
+      end
+
+      sig do
+        params(
+          node: SyntaxTree::Node,
+          position: Integer,
+        ).returns(T.nilable(SyntaxTree::TStringContent))
+      end
+      def find(node, position)
```

Comment:
We actually got rid of our previous implementation that used binary search in favour of `BaseRequest#locate`. It had the issue of not always returning the most specific node as possible.

To avoid duplication, can we please change `locate` to accept a target node (with its default being `nil`) and use it?

We can do something like
```ruby
possible_completion = locate(position, [SyntaxTree::Command, SyntaxTree::CallNode])
# perform argument and other checks
```

Reply:
Ah nice, I can change it to be that. Thanks for the pointer.

Reply:
It turns out `locate` wasn't accepting a list of node types to match, so I added it as an optional kwarg to the `locate` method and used it.